sopel/modules/safety.py

```diff
@@ -26,56 +27,107 @@
+    """Deprecated: Sets default_mode to "on"."""
+    default_mode = types.ValidatedAttribute("default_mode")
+    """Which mode to use in channels without a mode set."""
     known_good = types.ListAttribute('known_good')
     """List of "known good" domains to ignore."""
     vt_api_key = types.ValidatedAttribute('vt_api_key')
     """Optional VirusTotal API key (improves malicious URL detection)."""
+    domain_blocklist_url = types.ValidatedAttribute('domain_blocklist_url')
+    """Optional hosts-file formatted domain blocklist to use instead of StevenBlack's."""
 
 
-def configure(config):
+def configure(settings):
     """
     | name | example | purpose |
     | ---- | ------- | ------- |
-    | enabled\\_by\\_default | True | Enable URL safety in all channels where it isn't explicitly disabled. |
-    | known\\_good | sopel.chat,dftba.net | List of "known good" domains to ignore. |
+    | default\\_mode | on | Which mode to use in channels without a mode set. |
+    | known\\_good | sopel.chat,dftba.net | List of "known good" domains or regexes to ignore. |
     | vt\\_api\\_key | 0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef | Optional VirusTotal API key to improve malicious URL detection |
+    | domain\\_blocklist\\_url | https://example.com/bad-hosts.txt | Optional hosts-file formatted domain blocklist to use instead of StevenBlack's. |
     """
-    config.define_section('safety', SafetySection)
-    config.safety.configure_setting(
-        'enabled_by_default',
-        "Enable URL safety in channels that don't specifically disable it?",
+    settings.define_section("safety", SafetySection)
+    settings.safety.configure_setting(
+        "default_mode",
+        (
+            "Which mode should be used in channels that haven't specifically set one?"
+            "\n({})".format("/".join(SAFETY_MODES))
+        ),
+        default="on",
     )
-    config.safety.configure_setting(
+    settings.safety.configure_setting(
         'known_good',
-        'Enter any domains to whitelist',
+        "Enter any domains to allowlist",
     )
-    config.safety.configure_setting(
+    settings.safety.configure_setting(
         'vt_api_key',
         "Optionally, enter a VirusTotal API key to improve malicious URL "
-        "protection.\nOtherwise, only the StevenBlack list will be used."
+        "protection.\nOtherwise, only the configured hosts list will be used.",
+    )
+    settings.safety.configure_setting(
+        'domain_blocklist_url',
+        "Optionally, provide the URL for a hosts-file formatted domain "
+        "blocklist to use instead of StevenBlack's.",
     )
 
 
 def setup(bot):
-    bot.config.define_section('safety', SafetySection)
+    bot.settings.define_section("safety", SafetySection)
+
+    if bot.settings.safety.default_mode is None:
+        bot.settings.safety.default_mode = "on"
+    # migrate from enabled_by_default to default_mode
```

Comment:
I'm not sure what you're suggesting - If there was a way to use BooleanAttribute as optional I'd have done that and checked `is None` and `== False`, but if no default is specified we get `== False`. No way I saw to allow default_mode to override enabled_by_default without making the latter a pretend-bool.

Reply:
What I'm suggesting is to not bother removing `enabled_by_default` at all: keep it a boolean (with a default value set to `True`), and do the following:

* set `enabled_by_default`'s default value to `True`
* if `default_mode` is set, use that, and ignore `enabled_by_default`
* otherwise, if `enabled_by_default` is `True`, then set `default_mode` to `on` (or whatever appropriate value we want as the default)
* if it is set to `False`, then set `default_mode` to `off` or something similar to deactivate it

No need to transform `enabled_by_default` into a pseudo-boolean value with `None` for default, since you don't have to care about it. Silently do the migration, document it, and it'll be fine.

Reply:
Ah, so migrate then *save*? I was going to avoid an automatic save since last I saw that clobbers comments, formatting, ordering, etc. in the config file.

Reply:
Nah, no need to save. When people will upgrade to Sopel 9, having set `enabled_by_default` to `False`, they will get a warning that `enabled_by_default` is not a recognized setting. Given it's a major upgrade, people should be a bit more careful. The worst that can happen here is that they get safety activated by default instead. Not a big deal in my opinion.

Reply:
Ehh. I'm strongly against sending every URL the bot sees to VirusTotal by surprise, but I suppose if that's not wanted, they probably don't have a VT api key configured.
This could also be made more simple by being able to call BooleanAttribute.parse("on"), but eh.
See new push

Reply:
This new version is pretty good! And better than I suggested.